### PR TITLE
AggregationSelector is not needed anymore

### PR DESCRIPTION
**Pull request by fraillt**
Hello,

I'm not sure if this is a right way to approach this, but basically I think that `AggregationSelector` trait, and many `with_aggregation_selector` configuration methods on various public SDK types is not needed anymore, because newly introduced [Views](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#view) feature fully replaces this trait and removing `AggregationSelector` only makes things simpler/more understandable.

I created this PR as a proof to show that we can simply remove it without loosing any current functionality.

The only issue is this sentence in SDK:
> If the MeterProvider has no View registered, take the Instrument and apply the default Aggregation on the basis of instrument kind according to the [MetricReader](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#metricreader) instance’s aggregation property.

If I'm not missing anything, I think it could be changed to:
> If the MeterProvider has no View registered, take the Instrument and apply the [default aggregation](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#default-aggregation).

What do you think, do you agree, or I'm missing something important here?

## Changes

* Remove `AggregationSelector` trait
* Remove `with_aggregation_selector` configuration function from many public types.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
This looks okay to me, thanks. We need have CLA agreement signed (one-time) before accepting any changes to the repo, as this is the mandatory requirement from OpenTelemetry/Linux-Foundation organization which hosts the project.

Though Views allow one to change default aggregation, I haven't observed demand for that capability yet!. That capability is *not* exposed in OTel .NET Metrics, and no one has ever asked for it so far (its been GA since late 2021). We could even that capability from OTel Rust too for first stable release to keep public API surface minimal.

**Comment by fraillt** [comment]
Thanks for guidance and your input!

For my use case it was enough when `AggregatorSelector` had signature of `fn aggregator_for(&self, descriptor: &Descriptor) -> Option<Arc<dyn Aggregator + Send + Sync>>`. It wasn't super convenient to use, but having instrument `name` and `kind` was enough to set proper histogram bucket sizes.
`Views` are more powerful and I like to have a power, but in my opinion they are too powerful, especially considering that they are not requested that much :). Regarding .NET i'm a bit surprised, maybe they have non-compatible (with SDK) ways to achieve same things? (worst case you implement your own `MetricReader` :))

In my opinion, supporting 0-N number of views was a mistake, because it's not obvious how things behave without reading very carefully SDK (or even better source code), and what is worse, you'll get warnings if you use `*` to change default.
If it would be exactly 1 view, things could be much simpler. I would be happy to have a simple interface with a `fn change_view(&self, stream: StreamView) -> StreamView`.

* Default implementation would be `stream`
* Drop everything would be `stream.aggregation(Drop)`
* For more complicated scenarios `StreamView` could have some helper functions like `matches_name` that could use `*` and `?`, but for the most part you would write simple code.
```rust
if stream.matches_name("histogram*") {
  return stream.aggregation(Sum)
} 
if stream.scope.name =="mylib" {
  return stream.allowed_attributes([Key::new("size")])
} 
if stream.scope.name == "otherlib" && if stream.name == "my_metric" {
  return stream.description("even nicer description for other lib")
    .unit("X")
}
stream // leave unchanged default
// or .stream.aggregation(Drop) to drop the rest
```

Anyway, thanks again, and I'll try to get CLA signed.

**Comment by cijothomas** [comment]
> Regarding .NET i'm a bit surprised, maybe they have non-compatible (with SDK) ways to achieve same things? (worst case you implement your own MetricReader :))

No there is no other way. OTel .NET simply don't allow one to change aggregation. Ofcourse it does allow one to change HistogramBuckets, and also flip between Explicit vs Exponential Histogram, but thats about it. One cannot take a Counter and ask it to be aggregated as Histogram and things like this.

**Comment by cijothomas** [comment]
@fraillt Your concerns are valid. Views does have valid use cases, but it is also true that its somewhat easy to mis-use and cause issues. My original plan was to [remove Views completely from 1st stable release.](https://github.com/open-telemetry/opentelemetry-rust/issues/1719). It is entirely possible we do that, or possible that we trim it down.

The most common use case for Views is to change Histogram Buckets - @utpilla is looking at fixing https://github.com/open-telemetry/opentelemetry-rust/issues/1241 so one can change Histogram Buckets *without* having to learn about Views.

Is your scenario also just about customizing histogram buckets?

Irrespective of this, the direction of this PR is good.

**Comment by fraillt** [comment]
Yes, currently I need to customize histogram buckets, but I actually I might want to use rename or aggregation drop functionality as well... :)
Don't get me wrong, I actually think that views are very good addition, and given what they provide I think they have good design. It's just it might be too powerful and that power adds a costs in terms of user experience.

Anyway, what should I do to finish my work? Is updating appropriate CHANGELOG.md files would be enough?

**Comment by cijothomas** [comment]
_Anyway, what should I do to finish my work? Is updating appropriate CHANGELOG.md files would be enough?_

Please pick latest main, and see if the CI turns green. I'll get back to reviewing this after the https://github.com/open-telemetry/opentelemetry-rust/pull/2088 release is done! Thanks for waiting!

**Comment by fraillt** [comment]
This test is flaky.
```
 ---- prometheus_exporter_integration stdout ----
thread 'prometheus_exporter_integration' panicked at opentelemetry-prometheus/tests/integration_test.rs:412:5:
assertion `left == right` failed: sanitized attributes to labels
```
Do you know if anyone is working on it? What should I do?

File: opentelemetry-prometheus/Cargo.toml [hunk 1]
```diff
@@ -21,13 +21,13 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
 once_cell = { workspace = true }
-opentelemetry = { version = "0.24", default-features = false, features = ["metrics"] }
-opentelemetry_sdk = { version = "0.24", default-features = false, features = ["metrics"] }
+opentelemetry = { default-features = false, features = ["metrics"], path = "../opentelemetry" }
```
Comment (cijothomas) [inline opentelemetry-prometheus/Cargo.toml:24]: prometheus was intentionally excluded to use latest sdk to keep 1.0 scope minimized. It is okay to do another release with prometheus picking up newest api/sdk, but please do that into a separate PR, to keep PRs focused on one thing at time.

**Review by cijothomas — CHANGES_REQUESTED** [review]
The changes look good. Thank you for your contirbution!

Requesting changes just to remove the Prometheus exporter changes from this PR. Please submit that as a separate PR, and we can get it in, and do a release of Prometheus as well.

File: opentelemetry-sdk/src/metrics/reader.rs [hunk 2]
```diff
@@ -136,10 +117,9 @@ impl DefaultAggregationSelector {
     pub fn new() -> Self {
         Self::default()
     }
-}
 
-impl AggregationSelector for DefaultAggregationSelector {
-    fn aggregation(&self, kind: InstrumentKind) -> Aggregation {
+    /// Default aggregation for instrument kind
+    pub fn aggregation(&self, kind: InstrumentKind) -> Aggregation {
```
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/reader.rs:122]: We don't want this method to be public. In fact, since this is only used in `pipeline`, we could simply remove the struct `DefaultAggregationSelector` along with its methods and move this to `pipeline.rs` as a module level function similar to `aggregate_fn` and `is_aggregator_compatible`.

**Review by cijothomas — APPROVED** [review]
Thanks a lot for your contributions! Would love to see more contributions.

**Review by utpilla — APPROVED** [review]

**Comment by domenicquirl** [comment]
Hi folks, I just had a miserable time upgrading our `opentelemetry` versions because we make use of custom histogram buckets and after bumping `opentelemetry-otlp` the `AggregationSelector` was just gone. I'm still severely confused as to how I am now and will be in the future supposed to do this, so I wanted to point out a few things I ran into in hopes it helps improving them:

This change made in into the changelog for `opentelemetry-sdk` ([v0.26.0](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/CHANGELOG.md#v0260)), but there is nothing in the corresponding [entry](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-otlp/CHANGELOG.md#v0260) for `opentelemetry-otlp`, even though this PR removed `with_aggregation_selector` from the `OtlpMetricPipeline`. This should probably be listed as a breaking change.

The changelog entry that does exist for `opentelemetry-sdk` just says that the same functionality "should" be achievable by using views. It does not say whether this is actually possible, nor how one would do so.

Views seem to be under-documented in general. While https://github.com/open-telemetry/opentelemetry-rust/issues/1060 exists and examples were added for usage from `opentelemetry-sdk` in https://github.com/open-telemetry/opentelemetry-rust/pull/1170, I was not able to find any official documentation on how to use views at all with `opentelemetry-otlp`.  As far as I was able to find out, there is now way to actually do so directly through the `OtlpMetricsPipeline` returned from `opentelemetry_otlp::new_pipeline().metrics()`, which is how meter providers are constructed in the crate documentation. I did find https://github.com/open-telemetry/opentelemetry-rust/issues/1241, but was again left wondering how I was supposed to make use of the advisory in this context.

In the end, I think I made it work without using an `OtlpMetricsPipeline` by constructing exporter, reader and provider individually following https://github.com/open-telemetry/opentelemetry-rust/issues/1174#issuecomment-1660646295, which is also out of date already but pointed me in the right direction. 

When reading about views in https://github.com/open-telemetry/opentelemetry-rust/issues/1849, I stumbled across https://github.com/open-telemetry/opentelemetry-rust/issues/1849#issuecomment-2142966646, which seems to indicate that this method of customization will _stop working_ in upcoming releases, maybe being re-added in the future. You again suggest using https://github.com/open-telemetry/opentelemetry-rust/issues/1241 instead, but at this point I am just lost as to what plugs in where.

Apologies for highjacking this PR for my comments, but it does feel like the OTLP crate got the short end of the stick with these changes. Putting aside my confusion, imo at minimum a changelog entry should be added for the breaking change in `opentelemetry-otlp` and it would be nice if that showed or linked to how to replace an aggregation selector with a view or however else is supposed to be the "correct" way of customizing things here going forward.

**Comment by fraillt** [comment]
Hi,
I'm not deeply involved in all the details, but I think that Views are the only option at the moment... 
The reason why AggregationSelector was removed by this PR, is because at some point it became basically useless. It changed signature from `Descriptor -> Aggregator` to `InstrumentKind -> Aggregation`.
IMHO, views will not be removed, (because there's no replacement for them yet), and there are at least two people (you and me) who needs this functionality. :) at the same time I agree, its not trivial to use them, (I actually found it easier to just go source code and figure out how they work...) so I think there will still be some changes.

**Comment by utpilla** [comment]
@domenicquirl Apologies for the confusion! Since this was more of an API/SDK feature than OTLP exporter, we didn't update the Changelog for OTLP Exporter. However, as you pointed out, we could have done a better job at updating the SDK changelog.

The best way to do customize histogram buckets is through the `with_boundaries` Hint API. This API was added in #2135 and it is independent of the exporter being used. You can find the usage in the metrics-basic example: https://github.com/open-telemetry/opentelemetry-rust/blob/16c0e1034384f3897ccecc76b2aa1fb55f1f231a/examples/metrics-basic/src/main.rs#L86-L93

Just FYI Histogram Hint API only covers customization of buckets. Other aggregation customizations would still be covered by Views. We would be focusing on improving and fixing Views after the stable release.

**Comment by domenicquirl** [comment]
Thanks for following up! Very much appreciate your acknowledgement regarding the changelog.

> The best way to do customize histogram buckets is through the `with_boundaries` Hint API. This API was added in #2135 and it is independent of the exporter being used. You can find the usage in the metrics-basic example:

@utpilla I've seen both the hint / advisory API and the example by now, but I'm actually using `tracing_opentelemetry` with `opentelemetry_otlp`, so there's two more layers between my code and anything from `opentelemetry_sdk`. I can set up a `MeterProvider` which I pass to the tracing layer, but it is the [`MetricsLayer`](https://docs.rs/tracing-opentelemetry/latest/tracing_opentelemetry/struct.MetricsLayer.html) which is responsible for creating the histograms and which would thus need to pass boundaries for them. At the moment, it doesn't, and there isn't really a way to hook into its event processing that I know of that would allow me to do this.

It might be that this is fine and `tracing_opentelemetry` just needs time to catch up. There's an open issue on the repo (https://github.com/tokio-rs/tracing-opentelemetry/issues/108) about supporting custom histogram buckets using the advisory API, but it hasn't seen much action. 

That said, I think even for users of just `opentelemetry_otlp` without tracing it would be nice if there was a way to achieve this via the `OtlpMetricsPipeline`. It's reasonably more complicated having to drop down to the individual components and it's not just required for histograms: the pipeline currently doesn't offer any functionality for views (perhaps because of the temporary removal that was considered?).

**Comment by cijothomas** [comment]
@domenicquirl Thanks for taking time reporting the issue (very well written!).

Couple of follow ups:
1. The OtlpMetricsPipeline has a questionable future, depending on https://github.com/open-telemetry/opentelemetry-rust/issues/1810. Configuring SDK (i.e MeterProvider) should be done in an exporter-neutral way, but OtlpMetricsPipeline breaks that by putting too much responsibility to an exporter.
2. Can you share your use-cases of needing customizing AggregationSelector. We are quite likely to remove View from 1.0 scope, but want to understand common use-cases for needing Views. (The most common use case I am seeing is the histogram buckets, then cardinality capping/dropping-attributes.)
